### PR TITLE
feat(core): view engine - add debug information

### PR DESCRIPTION
**Pull request by tbosch**
Creates debug information for the renderer,
and also reports errors relative to the
declaration place in the template.

Part of #14013

### CODE REVIEW

File: modules/@angular/core/src/view/types.ts [hunk 1]
```diff
@@ -166,6 +162,7 @@ export interface ElementDef {
    * to indices in parent ElementDefs.
    */
   providerIndices: {[tokenKey: string]: number};
+  source: string;
```
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: Maybe call it `stack`?

File: modules/@angular/core/src/view/types.ts [hunk 2]
```diff
@@ -201,7 +199,10 @@ export interface ProviderOutputDef {
   eventName: string;
 }
 
-export interface TextDef { prefix: string; }
+export interface TextDef {
+  prefix: string;
+  source: string;
```
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: stack?

File: modules/@angular/core/src/view/types.ts [hunk 3]
```diff
@@ -361,4 +362,24 @@ export interface Services {
...
+// -------------------------------------
+export enum EntryAction {
+  CheckAndUpdate,
+  CheckNoChanges,
+  Create,
+  Destroy,
+  HandleEvent
+}
+
+export abstract class DebugContext extends RenderDebugInfo {
```
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: convert to interface with just properties

File: modules/@angular/core/src/view/element.ts [hunk 4]
```diff
@@ -18,6 +19,7 @@ export function anchorDef(
   if (matchedQueries) {
     matchedQueries.forEach(([queryId, valueType]) => { matchedQueryDefs[queryId] = valueType; });
   }
+  const source = isDevMode() ? sliceErrorStack(new Error().stack, 1, 2) : '';
```
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: stack

Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: Check for ie9, has no stack at all. Create one helper function that also includes `new Error().stack`.

File: modules/@angular/core/src/view/element.ts [hunk 5]
```diff
@@ -54,6 +56,7 @@ export function elementDef(
         ([BindingType.ElementClass, string] | [BindingType.ElementStyle, string, string] |
          [BindingType.ElementAttribute | BindingType.ElementProperty, string, SecurityContext])[],
     outputs?: (string | [string, string])[]): NodeDef {
+  const source = isDevMode() ? sliceErrorStack(new Error().stack, 1, 2) : '';
```
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: see above

File: modules/@angular/core/src/view/element.ts [hunk 6]
```diff
@@ -127,8 +130,11 @@ export function createElement(view: ViewData, renderHost: any, def: NodeDef): El
   const elDef = def.element;
   let el: any;
   if (view.renderer) {
-    el = elDef.name ? view.renderer.createElement(parentNode, elDef.name) :
-                      view.renderer.createTemplateAnchor(parentNode);
+    const debugContext = view.def.flags & ViewFlags.Debug ?
```
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: drop `ViewFlags.Debug`, always use `isDevMode`

File: modules/@angular/core/src/view/element.ts [hunk 7]
```diff
@@ -314,7 +324,7 @@ function setElementProperty(
   let renderValue = securityContext ? view.services.sanitize(securityContext, value) : value;
   if (view.renderer) {
     view.renderer.setElementProperty(renderNode, name, renderValue);
-    if (view.def.flags & ViewFlags.LogBindingUpdate) {
+    if (view.def.flags & ViewFlags.Debug & ViewFlags.DirectDom) {
```
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: should be `!DirectDom`

File: modules/@angular/core/src/view/provider.ts [hunk 8]
```diff
@@ -239,6 +251,17 @@ export function resolveDep(
...
 }
 
+export function createInjector(view: ViewData, elIndex: number): Injector {
+  return new Injector_(view, elIndex);
+}
+
+class Injector_ implements Injector {
+  constructor(private view: ViewData, private elIndex: number) {}
+  get(token: any, notFoundValue?: any): any {
+    return resolveDep(this.view, this.elIndex, {flags: 0, token: token, tokenKey: tokenKey(token)});
```
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: Use `DepFlags.None`

File: modules/@angular/core/src/view/provider.ts [hunk 9]
```diff
@@ -258,7 +281,7 @@ function checkAndUpdateProp(
     // so Closure Compiler will have renamed the property correctly already.
     provider[propName] = value;
 
-    if (view.def.flags & ViewFlags.LogBindingUpdate) {
+    if (view.def.flags & ViewFlags.Debug & ViewFlags.DirectDom) {
```
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: !DirectDom?

File: modules/@angular/core/src/view/text.ts [hunk 10]
```diff
@@ -6,12 +6,14 @@
...
+import {isDevMode} from '../application_ref';
 import {looseIdentical} from '../facade/lang';
 
-import {BindingDef, BindingType, NodeData, NodeDef, NodeFlags, NodeType, Services, TextData, ViewData, asElementData, asTextData} from './types';
-import {checkAndUpdateBinding} from './util';
+import {BindingDef, BindingType, DebugContext, NodeData, NodeDef, NodeFlags, NodeType, Services, TextData, ViewData, ViewFlags, asElementData, asTextData} from './types';
+import {checkAndUpdateBinding, sliceErrorStack} from './util';
 
 export function textDef(constants: string[]): NodeDef {
+  const source = isDevMode() ? sliceErrorStack(new Error().stack, 1, 2) : '';
```
Comment (tbosch) [inline modules/@angular/core/src/view/text.ts]: See comment at element.ts

File: modules/@angular/core/src/view/util.ts [hunk 11]
```diff
@@ -76,4 +78,94 @@ export function renderNode(view: ViewData, def: NodeDef): any {
...
+    // Chrome adds a first line with the error message
+    row++;
+  }
+  return lines[row];
+}
+
+export function resolveViewDefinition(factory: ViewDefinitionFactory): ViewDefinition {
+  let value: ViewDefinition = (<any>factory)[VIEW_DEFINITION_CACHE_KEY];
+  if (!value) {
+    const currentStackLines = new Set(new Error().stack.split('\n'));
```
Comment (tbosch) [inline modules/@angular/core/src/view/util.ts]: remove this line!

File: modules/@angular/core/src/view/util.ts [hunk 12]
```diff
@@ -76,4 +78,94 @@ export function renderNode(view: ViewData, def: NodeDef): any {
     case NodeType.Text:
       return asTextData(view, def.index).renderText;
   }
-}
\ No newline at end of file
+}
+
+const VIEW_DEFINITION_CACHE_KEY = '$$cache';
+
+export function getStackEntry(stack: string, row: number) {
```
Comment (tbosch) [inline modules/@angular/core/src/view/util.ts]: remove

File: modules/@angular/core/src/view/view.ts [hunk 13]
```diff
@@ -31,7 +32,10 @@ export function viewDef(
   const reverseChildNodes: NodeDef[] = new Array(nodesWithoutIndices.length);
   let viewBindingCount = 0;
   let viewDisposableCount = 0;
-  let viewFlags = 0;
+  let viewNodeFlags = 0;
+  if (isDevMode()) {
+    flags |= ViewFlags.Debug;
```
Comment (tbosch) [inline modules/@angular/core/src/view/view.ts]: maybe remove this flag?

File: modules/@angular/core/src/view/view.ts [hunk 14]
```diff
@@ -256,14 +268,32 @@ function createView(
...
   view.component = component;
   view.context = context;
+}
+
+const createViewNodes: (view: ViewData) => void =
+    entryAction(EntryAction.CheckNoChanges, _createViewNodes);
+
+function _createViewNodes(view: ViewData) {
+  let renderHost: any;
+  if (view.parent && view.component === view.context) {
```
Comment (tbosch) [inline modules/@angular/core/src/view/view.ts]: create helper function for this, also used in debugContext...

File: modules/@angular/core/src/view/view.ts [hunk 15]
```diff
@@ -276,9 +306,13 @@ function initView(view: ViewData, renderHost: any, component: any, context: any)
         if (nodeDef.provider.component) {
           const hostElIndex = nodeDef.parent;
           componentView = createView(
-              view.services, view, hostElIndex, hostElIndex, nodeDef.provider.component());
+              view.services, view, hostElIndex, hostElIndex,
+              resolveViewDefinition(nodeDef.provider.component));
```
Comment (tbosch) [inline modules/@angular/core/src/view/view.ts]: Maybe use a WeakMap to cache ViewDefinitions?

File: modules/@angular/core/test/view/binding_spec.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+import {RenderComponentType, RootRenderer, Sanitizer, SecurityContext, ViewEncapsulation, getDebugNode} from '@angular/core';
+import {DefaultServices, NodeDef, NodeFlags, Services, ViewData, ViewDefinition, ViewFlags, ViewHandleEventFn, ViewUpdateFn, anchorDef, checkAndUpdateView, checkNoChangesView, checkNodeDynamic, checkNodeInline, createRootView, elementDef, rootRenderNodes, setCurrentNode, textDef, viewDef} from '@angular/core/src/view/index';
+import {inject} from '@angular/core/testing';
+import {getDOM} from '@angular/platform-browser/src/dom/dom_adapter';
+
+import {isBrowser, setupAndCheckRenderer} from './helper';
+
+export function main() {
```
Comment (tbosch) [inline modules/@angular/core/test/view/binding_spec.ts]: remove this file

File: modules/@angular/core/test/view/element_spec.ts [hunk 17]
```diff
@@ -121,7 +104,10 @@ function defineTests(config: {directDom: boolean, viewFlags: number}) {
                       [BindingType.ElementProperty, 'value', SecurityContext.NONE]
                     ]),
               ],
-              config.update));
+              (view) => {
+                setCurrentNode(view, 0);
+                checkNodeInlineDynamic(inlineDynamic, ['v1', 'v2']);
```
Comment (tbosch) [inline modules/@angular/core/test/view/element_spec.ts]: rename: checkNodeInlineOrDynamic

File: modules/@angular/core/test/view/element_spec.ts [hunk 18]
```diff
@@ -346,6 +317,22 @@ function defineTests(config: {directDom: boolean, viewFlags: number}) {
           expect(preventDefaultSpy).toHaveBeenCalled();
         });
 
+        it('should report debug info on output errors', () => {
```
Comment (tbosch) [inline modules/@angular/core/test/view/element_spec.ts]: event errors

File: modules/@angular/core/test/view/element_spec.ts [hunk 19]
```diff
@@ -346,6 +317,22 @@ function defineTests(config: {directDom: boolean, viewFlags: number}) {
...
+
+          let err: any;
+          try {
+            addListenerSpy.calls.mostRecent().args[1]('SomeEvent');
+          } catch (e) {
+            err = e;
+          }
+          expect(err).toBeTruthy();
+          expect(err.message).toBe('Test');
+          expect(err.context.renderNode).toBe(asElementData(view, 0).renderElement);
```
Comment (tbosch) [inline modules/@angular/core/test/view/element_spec.ts]: assert `view` and `nodeIndex`

File: modules/@angular/core/test/view/services_spec.ts [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+              NodeFlags.None, null, 0, AComp, [], null, null,
+              () => compViewDef([
+                elementDef(NodeFlags.None, [['#ref', QueryValueType.ElementRef]], 2, 'span'),
+                providerDef(NodeFlags.None, null, 0, AService, []), textDef(['a'])
+              ])),
+        ]));
+        return view;
+      }
+
+      it('should provider data for elements', () => {
```
Comment (tbosch) [inline modules/@angular/core/test/view/services_spec.ts]: should provide...

**Comment by tbosch** [comment]
Add tests for `setBindingDebugInfo`.

Make all changes in a separate commit and ping @vicb

File: modules/@angular/core/src/view/types.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: Done

File: modules/@angular/core/src/view/types.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: Done

File: modules/@angular/core/src/view/types.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: Done

File: modules/@angular/core/src/view/element.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: Done

File: modules/@angular/core/src/view/element.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: Done

File: modules/@angular/core/src/view/element.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: Done

File: modules/@angular/core/src/view/element.ts (hunk 7 above)
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: Done

File: modules/@angular/core/src/view/provider.ts (hunk 8 above)
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: Done

File: modules/@angular/core/src/view/provider.ts (hunk 9 above)
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: Done

File: modules/@angular/core/src/view/text.ts (hunk 10 above)
Comment (tbosch) [inline modules/@angular/core/src/view/text.ts]: Done

File: modules/@angular/core/src/view/util.ts (hunk 11 above)
Comment (tbosch) [inline modules/@angular/core/src/view/util.ts]: Done

File: modules/@angular/core/src/view/util.ts (hunk 12 above)
Comment (tbosch) [inline modules/@angular/core/src/view/util.ts]: Done

File: modules/@angular/core/src/view/view.ts (hunk 13 above)
Comment (tbosch) [inline modules/@angular/core/src/view/view.ts]: Done

File: modules/@angular/core/src/view/view.ts (hunk 14 above)
Comment (tbosch) [inline modules/@angular/core/src/view/view.ts]: Done

File: modules/@angular/core/src/view/view.ts (hunk 15 above)
Comment (tbosch) [inline modules/@angular/core/src/view/view.ts]: Done

File: modules/@angular/core/test/view/binding_spec.ts (hunk 16 above)
Comment (tbosch) [inline modules/@angular/core/test/view/binding_spec.ts]: Done

File: modules/@angular/core/test/view/element_spec.ts (hunk 17 above)
Comment (tbosch) [inline modules/@angular/core/test/view/element_spec.ts]: Done

File: modules/@angular/core/test/view/element_spec.ts (hunk 18 above)
Comment (tbosch) [inline modules/@angular/core/test/view/element_spec.ts]: Done

File: modules/@angular/core/test/view/element_spec.ts (hunk 19 above)
Comment (tbosch) [inline modules/@angular/core/test/view/element_spec.ts]: Done

File: modules/@angular/core/test/view/services_spec.ts (hunk 20 above)
Comment (tbosch) [inline modules/@angular/core/test/view/services_spec.ts]: Done

**Comment by tbosch** [comment]
Added the missing tests.

File: modules/@angular/core/src/view/element.ts [hunk 21]
```diff
@@ -19,7 +19,8 @@ export function anchorDef(
   if (matchedQueries) {
     matchedQueries.forEach(([queryId, valueType]) => { matchedQueryDefs[queryId] = valueType; });
   }
-  const source = isDevMode() ? sliceErrorStack(new Error().stack, 1, 2) : '';
+  // skip the call to sliceErrorStack itself + the call to this function.
+  const stack = isDevMode() ? sliceErrorStack(2, 3) : '';
```
Comment (vicb) [inline modules/@angular/core/src/view/element.ts]: stack -> caller ?

**Comment by vicb** [comment]
LGTM, cleanup ?

**Comment by tbosch** [comment]
Will do...

**Review by vicb — APPROVED** [review]

File: modules/@angular/core/src/view/element.ts (hunk 21 above)
Comment (tbosch) [inline modules/@angular/core/src/view/element.ts]: I think I'd rather switch back to `source` again, as neither `stack` nor `caller` convey the meaning of `the position in the template`.

**Comment by tbosch** [comment]
@mhevery This marker code-review/pullapprove wrong (@vicb approved it) , @IgorMinar will look into it...

**Review by mhevery — APPROVED** [review]

**Pull request closed without merging**